### PR TITLE
`[digital-carbon]` c3 async retires for J-credit and ECO

### PR DESCRIPTION
**Pull request by psparacino**
Adds support for async retires of C3 J-credit and EcoRegistry tokens.

also adds support for amoy.

Currently using a [blockHandler](https://github.com/KlimaDAO/klima-subgraph/pull/151/files) at ~2min intervals to get the tokenURI if the tokenURI isn't found in the initial contract call. In some cases it appears the tokenURI isn't set when the actual retirement takes place.

### CODE REVIEW

File: lib/abis/C3ProjectTokenFactory.json
Comment (0xAurelius) [inline lib/abis/C3ProjectTokenFactory.json:897]: Are these new ABIs strictly a superset? So no old functionality will break from swapping these out?

File: polygon-digital-carbon/src/KlimaAggregator.ts
Comment (0xAurelius) [inline polygon-digital-carbon/src/KlimaAggregator.ts:10]: I don't see this referenced anywhere in this file, only the subsequent one down below?

File: polygon-digital-carbon/schema.graphql
Comment (0xAurelius) [inline polygon-digital-carbon/schema.graphql:388]: Should we not call this a `Retire` request? Or does C3's own language in the contracts call it "offset" as the verb?

File: polygon-digital-carbon/src/RetirementHandler.ts
Comment (0xAurelius) [inline polygon-digital-carbon/src/RetirementHandler.ts:110]: Isn't this going to get out of hand with PURO or other registries coming? Maybe should be a list we check against?

File: polygon-digital-carbon/src/RetirementHandler.ts
Comment (0xAurelius) [inline polygon-digital-carbon/src/RetirementHandler.ts:251]: again with the "offset" language - don't love it

File: polygon-digital-carbon/src/RetirementHandler.ts
Comment (0xAurelius) [inline polygon-digital-carbon/src/RetirementHandler.ts:295]: Does this need to be implemented before merging? Seems important to avoid double counting...

File: polygon-digital-carbon/src/TransferHandler.ts
Comment (0xAurelius) [inline polygon-digital-carbon/src/TransferHandler.ts:1]: should this still be imported? seems unused

File: polygon-digital-carbon/src/utils/C3.ts
Comment (0xAurelius) [inline polygon-digital-carbon/src/utils/C3.ts:9]: best practice for strings like this is to store them in a constant and then always reference it like `AWAITING`

not a big deal, just a nit pick

File: polygon-digital-carbon/subgraph.template.yaml
Comment (0xAurelius) [inline polygon-digital-carbon/subgraph.template.yaml:6]: could you throw in a comment about how this grafting block should/shouldn't be used?

**Review by 0xAurelius — CHANGES_REQUESTED** [review]
A few nit picks, one general concern about use of the term "offset" - want to make sure it's consistent with how smart contracts are written and not just willy nilly since we've tried to be deliberate about coalescing on "credit" / "project"

File: lib/abis/C3ProjectTokenFactory.json
Comment (psparacino) [reply lib/abis/C3ProjectTokenFactory.json:897]: Yes, the new abis just add the async events & functions

File: polygon-digital-carbon/src/KlimaAggregator.ts
Comment (psparacino) [reply polygon-digital-carbon/src/KlimaAggregator.ts:10]: yes, an old import; removed

File: polygon-digital-carbon/schema.graphql
Comment (psparacino) [reply polygon-digital-carbon/schema.graphql:388]: in the C3 contracts the function is `offsetFor` which is why I initially named it this way. Changed to Retire though to stay consistent on the Carbonmark side.

File: polygon-digital-carbon/src/RetirementHandler.ts
Comment (psparacino) [reply polygon-digital-carbon/src/RetirementHandler.ts:295]: Ended up not need this with the current implementation, removed.

File: polygon-digital-carbon/src/RetirementHandler.ts
Comment (psparacino) [reply polygon-digital-carbon/src/RetirementHandler.ts:110]: This event is only for C3 async retires, which are currently only JCS and ECO. So Puro shouldn't ever interact with this handler. If there continue to be more registries with async retirements from the C3ProjectTokenFactory though then yes a more organized list would avoid confusion. This is the only event that currently has a conditional between async and non-async c3 retirements.

File: polygon-digital-carbon/src/utils/C3.ts
Comment (psparacino) [reply polygon-digital-carbon/src/utils/C3.ts]: created an assemblyscript enum. The enums from the actual schema aren't generated via codegen so I don't see an immediate way to mirror the enum but this at least creates more visibility & a single source to update if the schema enum changes.

**Review by 0xAurelius — APPROVED** [review]
LGTM but defer to @cujowolf on final approval

**Review by cujowolf — APPROVED** [review]
LGTM

**Pull request merged**
